API_MANIFESTO.md

```diff
@@ -0,0 +1,91 @@
+# The API Manifesto
+
+# Infinispan API dependency module
+
+The ```infinispan-api``` dependency will contain the public API. This module will be independent from the other modules. 
+Dependency with implementations will be **logical**. This will keep the modularity needed for the different use cases.
+
+Examples:
+ * Cache use case
+ 
+ * NoSQL data store use case
+ 
+ * Tasks, concurrency, counters uses cases
+
+The dependencies to thirty-party libraries must be avoided. This module should contain dependencies to JDK classes, 
+not more.
+
+# User First
+
+* Before making an API public, consider to test the usability with someone not familiar with the feature
+
+* Convention over Configuration: Make sure a default behaviour exists. Don't make me think.
+
+* Avoid exposing internals as much as possible. Even if your feature lays on top of a cache today, internals can change.
+
+* Each public API/Feature must have a simple tutorial showcasing the example. Without a demo, the API does not exist
+
+* If you are unsure about an API, mark it as ```@Experimental```. 
+
+* Don't let ```@Experimental``` tag last forever. 
```

Comment:
Define !forever... Should @Experimental become supported API in next major?

Reply:
I'm not sure about what forever should be. I was thinking like you, like next mayor or two mayor versions. And if the API does not exit experimental, is it worth to maintain it/have it?

Reply:
I'm not sure about what forever should be. I was thinking like you, like next mayor or two mayor versions. And if the API does not exit experimental, is it worth to maintain it/have it?

Reply:
I'm not sure about letting an API be experimental forever. For example, in the case of the functional map API, we've seen it's more useful for our own use cases than external use cases, so you could just move it and make it internal only. If enough people demand it it can be made non-experimental? Maybe let the users drive that?

Reply:
Ok, let's treat it case by case and user driven

Reply:
The more I think about it, the more I'm coming to the conclusion that this should be an exception and that we should not have @Experimental API in our public API. So once we consider that the API is worth to be public, really public and not used by a minority, then it should be promoted to the public API. Again, this should be used as an exception.
Or maybe not only having the @Experimental annotation, but a dedicated package? 
WDYT @tristantarrant ?